### PR TITLE
CTFE engine refactor

### PR DESCRIPTION
**Pull request by RalfJung**
* Value gets renamed to `Operand`, so that now `interpret::{Place, Operand}` are the "dynamic" versions of `mir::{Place, Operand}`.
* `Operand` and `Place` share the data for their "stuff is in memory"-base in a new type, `MemPlace`. This also makes it possible to give some more precise types in other areas. Both `Operand` and `MemPlace` have methods available to project into fields (and other kinds of projections) without causing further allocations.
* The type for "a `Scalar` or a `ScalarPair`" is called `Value`, and again used to give some more precise types.
* All of these have versions with an attached layout, so that we can more often drag the layout along instead of recomputing it. This lets us get rid of `PlaceExtra::Downcast`. `MPlaceTy` and `PlaceTy` can only be constructed in place.rs, making sure the layout is handled properly. (The same should eventually be done for `ValTy` and `OpTy`.)
 This is used to check, when copying an operand to a place, that the sizes match (which caught a bunch of bugs).
* All the high-level functions to write typed memory take a `Place`, and live in `place.rs`. All the high-level typed functions to read typed memory take an `Operand`, and live in `operands.rs`.
* Remove `cur_frame` and handling of signedess from memory (catching a bug in the float casting code).
* [Only functional change] Enable sanity check to recurse below dyn traits and slices.

r? @oli-obk 

Cc @eddyb

### CODE REVIEW

**Comment by RalfJung** [comment]
We'll want a perf run.

@bors try

**Comment by RalfJung** [comment]
Hu? I rebased right before I submitted this...

File: src/librustc_mir/interpret/place.rs [hunk 1]
```diff
@@ -1,472 +1,755 @@
...
+}
+
+// Validation needs to hash MPlaceTy, but we cannot hash Layout -- so we just hash the type
+impl<'tcx> Hash for MPlaceTy<'tcx> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.mplace.hash(state);
+        self.layout.ty.hash(state);
+    }
+}
+impl<'tcx> PartialEq for MPlaceTy<'tcx> {
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]: can't this be derived?

File: src/librustc_mir/interpret/place.rs [hunk 2]
```diff
@@ -1,472 +1,755 @@
...
+    }
 
-        // Do not allocate in trivial cases
-        let (base_ptr, base_align, base_extra) = match base {
-            Place::Ptr { ptr, align, extra } => (ptr, align, extra),
+    /// Write a value to a place
+    pub fn write_value(
+        &mut self,
+        src_val: Value,
+        dest : PlaceTy<'tcx>,
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]: space after `dest`

File: src/test/ui/union-ub-fat-ptr.rs [hunk 3]
```diff
@@ -66,16 +66,18 @@ trait Trait {}
...
 
 // OK
 const A2: &[u8] = unsafe { SliceTransmute { repr: SliceRepr { ptr: &42, len: 1 } }.slice};
-// should lint
+// bad
 const B2: &[u8] = unsafe { SliceTransmute { repr: SliceRepr { ptr: &42, len: 999 } }.slice};
+//~^ ERROR this constant likely exhibits undefined behavior
 // bad
 const C2: &[u8] = unsafe { SliceTransmute { bad: BadSliceRepr { ptr: &42, len: &3 } }.slice};
 //~^ ERROR this constant likely exhibits undefined behavior
```
Comment (oli-obk) [inline src/test/ui/union-ub-fat-ptr.rs]: can you also add a test where a trait object has the correct vtable and pointer but the object is wrong? So create the object unsafely with brokenness and then downcast

**Review by oli-obk — CHANGES_REQUESTED** [review]
This refactoring was very overdue. Thanks so much for doing it. So many great things in there.

**Comment by RalfJung** [comment]
Okay we don't need a perf run yet... triggered by a warning that some test takes >60s, I benchmarked `src/test/run-pass/mir_heavy_promoted.rs` and it's gotten a *lot* slower. :/

**Comment by RalfJung** [comment]
Uh. Perf says

> -   72,91%     0,00%  rustc     librustc_lint-c3cbb903fb8aa258.so      [.] rustc_lint::builtin::check_const

However, I also have debug-assertions enabled, so let's see if that makes a difference.

**Comment by RalfJung** [comment]
After compiling without debug assertions, it's still 50% slower.

File: src/librustc_mir/interpret/place.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: No, `TyLayout` has no `PartialEq`. So we just compare the type.

File: src/librustc_mir/interpret/operand.rs [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+            Value::ScalarPair(..) => bug!("got ScalarPair for type: {:?}", op.layout.ty),
+            Value::Scalar(val) => Ok(val),
+        }
+    }
+
+    pub fn uninit_operand(&mut self, layout: TyLayout<'tcx>) -> EvalResult<'tcx, Operand> {
+        // This decides which types we will use the Immediate optimization for, and hence should
+        // match what `try_read_value` and `eval_place_to_op` support.
+        if layout.is_zst() {
+            return Ok(Operand::Immediate(Value::Scalar(ScalarMaybeUndef::Undef)));
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: Well, why not?

File: src/librustc_mir/interpret/place.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: Fixed.

File: src/librustc_mir/interpret/operand.rs [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+        mplace: MPlaceTy<'tcx>,
+    ) -> EvalResult<'tcx, Option<Value>> {
+        if mplace.extra != PlaceExtra::None {
+            return Ok(None);
+        }
+        let (ptr, ptr_align) = mplace.to_scalar_ptr_align();
+        self.memory.check_align(ptr, ptr_align)?;
+
+        if mplace.layout.size.bytes() == 0 {
+            return Ok(Some(Value::Scalar(ScalarMaybeUndef::Scalar(Scalar::Bits { bits: 0, size: 0 }))));
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: Done.

File: src/librustc_mir/interpret/place.rs [hunk 6]
```diff
@@ -1,472 +1,755 @@
...
+    }
+
+    #[inline]
+    pub(super) fn len(self) -> u64 {
+        // Sanity check
+        let ty_len = match self.layout.fields {
+            layout::FieldPlacement::Array { count, .. } => count,
+            _ => bug!("Length for non-array layout {:?} requested", self.layout),
+        };
+        if let PlaceExtra::Length(len) = self.extra {
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: I was thinking so... @eddyb should know.^^ What is the layout of `[u8]`?

I seem to remember it is an "array" of length 0.

File: src/test/ui/union-ub-fat-ptr.rs (hunk 3 above)
Comment (RalfJung) [reply src/test/ui/union-ub-fat-ptr.rs]: I added these:

```rust
// bad data *inside* the trait object
const G: &Trait = &unsafe { BoolTransmute { val: 3 }.bl };

// bad data *inside* the slice
const H: &[bool] = &[unsafe { BoolTransmute { val: 3 }.bl }];
```
Does that look like it is testing the right thing?

Comment (oli-obk) [reply src/test/ui/union-ub-fat-ptr.rs]: You forgot to push

Comment (RalfJung) [reply src/test/ui/union-ub-fat-ptr.rs]: No I just waited for `--bless` to finish before pushing. ;)

File: src/librustc_mir/interpret/operand.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: Okay I will just use `Scalar::zst()`

[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc_mir/interpret/operand.rs]: What made you change your mind?

[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: Talking with @eddyb :D And also thinking about what this even means... I mean, ZSTs don't really have a value. The read functions short-circuit them before we even hit memory, and they always return `Scalar::zst()`. So this seems more consistent.

Also, 0 bytes of unintiailized memory are not really `undef`. They are just empty. As in, if you think of a value as a `&[Byte]`, where `Byte` is either `u8` or `Undef` -- then a value of size 0 cannot be `Undef`.

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
r=me w/o perf regressions

**Comment by RalfJung** [comment]
I guess we do not have to wait with the crater run until I completed the perf work...

@bors try

**Comment by RalfJung** [comment]
Tests looking good locally.

@bors try

**Comment by RalfJung** [comment]
@rust-lang/infra could we get a crater run? This makes the sanity check on constants stronger, and also touches enough of the CTFE engine to potentially fail somewhere.

**Comment by RalfJung** [comment]
All right, and this more than fixes the perf issue -- in my local benchmark, the big array case is now 18% *faster* than previously (from 11.9s to 10.1s). I could still shove off a few additional percent at the expense of code reuse and clarity, but I guess this is good enough for now.

For a perf run:
@bors try
(Sorry for all the try runs, I did not think I would finish the perf work today)

**Comment by kennytm** [comment]
@rust-timer build 84472da909f0698a692c0acea9a4b09cbf5471f1

**Comment by Valinora** [comment]
Perf results are in

**Comment by RalfJung** [comment]
Well that's what I would call "mixed results" - from 35% faster to 25% slower...

**Comment by oli-obk** [comment]
Slice and unsize checks are expected to be slower since they are actually checked now. for all the others we'd need to check the code for possible new checks that are happening

**Comment by RalfJung** [comment]
Stuff is not (just) slower during checks though, but also during actual computation.

**Comment by pietroalbini** [comment]
@craterbot run start=master#d5b6b95aef94169b5dbe4dbb1357d4bab1fc9800 end=try#84472da909f0698a692c0acea9a4b09cbf5471f1 mode=check-only

**Comment by RalfJung** [comment]
I optimized `push_stack_frame` to be able to iterate over the locals, instead of accessing them by index all the time. This helps a lot locally.

@bors try

**Comment by Mark-Simulacrum** [comment]
@rust-timer build 348ad7c68167cc680b13f2b4c835d5f00e4265d0

**Comment by oli-obk** [comment]
Looking at the MIR of arrays initialized with const fn call results, it looks to me like there are a lot of useless assignments and locals. Maybe assignment by value got slower?

**Comment by RalfJung** [comment]
> Maybe assignment by value got slower?

Indeed, `Use` got slower -- it now computes the layout twice instead of once. Turns out several million extra hits of the query cache show up here.

I fixed that now, and the benchmark is now 17% *faster* than `master` in my local measurements :)

@bors try

**Comment by kennytm** [comment]
@rust-timer build 86419f981b1097376824c441e716361ca9ccd558

**Comment by memoryruins** [comment]
Perf is ready ^^

**Comment by RalfJung** [comment]
I cannot reproduce any significant difference on a full build of coercions.

Also, that benchmark consists of a huge `[&'static str; 0xffff]`. Validation now recurses below `&str`, which could explain the difference (in particular, I think validation happens even on a clean incremental build because lints are run?). I could pack out this part just to make sure that there is no other regression. @oli-obk?

**Comment by oli-obk** [comment]
I think the regression is fine. I'll create an issue about figuring out how to make incremental not run validation twice

**Comment by RalfJung** [comment]
It only runs once -- and incremental is supposed to show all diagnostics again, right?

Or is the issue about caching diagnostics instead of recomputing them? Doesn't that happen automatically for all lints?

**Comment by RalfJung** [comment]
I *finally* reproduce the `coercions` perf regression locally, and I can confirm that after removing the recursion down below `&str`, the 7% slowdown (that I am seeing locally, perf sees 9%) turns into a 9% *speedup*.

So I am confident now that the remaining regressions are a duplicate of #52849.

**Comment by RalfJung** [comment]
The one "fixed" reported by crateer is spurious.

One regression is legit:
```
Unexpected fat pointer target type TyAdt(c_utf8::CUtf8, [])
```

The other seems spurious
```
Aug 21 02:09:09.857 INFO kablam! thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 30, kind: Other, message: "Read-only file system" }', libcore/result.rs:983:5
```

**Comment by RalfJung** [comment]
Fixed the regression that crater found.

@oli-obk I think this is good to go :)

**Comment by RalfJung** [comment]
Btw, even though this updates miri, it will not actually make miri green until https://github.com/rust-lang/rust/pull/53536 lands. Would bors reject this PR if miri does not turn green?

**Comment by oli-obk** [comment]
This now does not update the submodule anymore. The submodule update will come in a separate PR.

@bors r+

**Comment by RalfJung** [comment]
I think this PR is way too big to be backported.

If need be we can disable validation on beta.

**Comment by pnkfelix** [comment]
in the discussion on #52849, it sounds like one of the big performance issues came from hashing the memory state, which was then something that the discussion talked about removing. Was hashing the memory state actually removed (I cannot tell from skimming the titles of the commit series) on this PR (#53424)? And if it was the case, can that part of the diff be backported on its own?

**Comment by RalfJung** [comment]
No it has not been removed here, but it will be removed in https://github.com/rust-lang/rust/pull/52626.

And yes, just removing it from the hash is trivial to backport. Do you want me to do that?

**Comment by pnkfelix** [comment]
@RalfJung well, that depends: does removing the memory from the hash indeed resolve the majority of #52849, without causing any known regressions elsewhere?

**Comment by RalfJung** [comment]
Yes and no. The majority of that issue is fully intentional -- we walk the resulting value computed for every constant, in benchmarks that consists only of a single constant, that has a huge impact.

Two unintentional changes that are also part of the same issue (hence the rather messy overall situation) are:
1. Computing every constant twice, which is fixed by https://github.com/rust-lang/rust/pull/52925. This is "beta-nominated" -- does that mean it hasn't landed in beta yet?
2. A strange effect occurring in two artificial benchmarks where the loop detector goes crazy and stops compilation to a halt. We don't know yet why this only started with the sanity check, but the fix is to stop hashing memory. We do not know of any real code affected by this.

File: src/librustc_mir/interpret/place.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_mir/interpret/place.rs]: You might want to include a few more things in the Eq/Hash, such as an `Option<usize>` where `Some(index)` means `Variants::Single { index }`, while `None` is for all other `Variants`.

**Pull request merged**